### PR TITLE
SDRS 2 -- /status endpoint enhancement and additional validation for creating rule

### PR DESCRIPTION
**Pull request by eshen1991**
**/status endpoint enhancement**

- check DB connection when app starts

- provide app status on web server, DB, and Pub/Sub

**Additional validation when creating rule**

- Check if the bucket exist in GCS

- Check if non-nesting rule is violated

### CODE REVIEW

File: src/main/java/com/google/gcs/sdrs/SdrsApplication.java [hunk 1]
```diff
@@ -165,9 +175,17 @@ private static String getPropertyValueFromEnv(String token) {
     return System.getenv(envVariable);
   }
 
-  private static void registerPubSub() {
+  private static boolean registerPubSub() {
```
Comment (salguerod) [inline src/main/java/com/google/gcs/sdrs/SdrsApplication.java]: For the change to a  boolean return type from void consider changing the method name to "is" as is the convention for conveying clear readability

File: src/main/java/com/google/gcs/sdrs/service/impl/RetentionRulesServiceImpl.java [hunk 2]
```diff
@@ -67,21 +71,31 @@ public RetentionRulesServiceImpl() {
...
-            rule.getProjectId(), rule.getDataStorageName(), true, rule.getRetentionRuleType());
+    List<RetentionRule> bucketRules =
+        ruleDao.findRulesByStorageRoot(
+            rule.getProjectId(),
+            RetentionUtil.getBucketName(rule.getDataStorageName()),
+            rule.getRetentionRuleType(),
+            true);
 
-    RetentionRule newRule;
+    RetentionRule existing = getExistingRule(bucketRules, rule);
```
Comment (salguerod) [inline src/main/java/com/google/gcs/sdrs/service/impl/RetentionRulesServiceImpl.java]: consider renaming to existingRule for increased readability

File: src/main/java/com/google/gcs/sdrs/controller/ServiceLayerException.java [hunk 3]
```diff
@@ -0,0 +1,50 @@
+/*
```
Comment (salguerod) [inline src/main/java/com/google/gcs/sdrs/controller/ServiceLayerException.java]: is this new class necessary ? can we just add the string instance variable to the extended base class and inject the appropriate message upon object instantiation as an argument?

File: src/main/java/com/google/gcs/sdrs/dao/impl/RetentionRuleDaoImpl.java [hunk 4]
```diff
@@ -206,6 +206,32 @@ public RetentionRule findGlobalRuleByProjectId(String projectId) {
     return result;
   }
 
+  @Override
+  public List<RetentionRule> findRulesByStorageRoot(
```
Comment (salguerod) [inline src/main/java/com/google/gcs/sdrs/dao/impl/RetentionRuleDaoImpl.java]: what is a storageRoot - the bucket or a particular "path" prefix?  Either way can we add that in the method Javadoc?

File: src/main/java/com/google/gcs/sdrs/service/impl/RetentionRulesServiceImpl.java [hunk 5]
```diff
@@ -90,19 +104,39 @@ public Integer createRetentionRule(RetentionRuleCreateRequest rule, UserInfo use
...
     }
 
     return newRule.getId();
   }
 
+  private RetentionRule getExistingRule(
+      List<RetentionRule> rules, RetentionRuleCreateRequest ruleRequest) throws SQLException {
+    if (rules != null) {
+      for (RetentionRule rule : rules) {
+        String dataStroargeName = ruleRequest.getDataStorageName();
```
Comment (salguerod) [inline src/main/java/com/google/gcs/sdrs/service/impl/RetentionRulesServiceImpl.java]: dataStroargeName to dataStorageName

File: src/main/java/com/google/gcs/sdrs/SdrsApplication.java (hunk 1 above)
Comment (eshen1991) [reply src/main/java/com/google/gcs/sdrs/SdrsApplication.java]: change it to void and log info inside of the method

File: src/main/java/com/google/gcs/sdrs/controller/ServiceLayerException.java (hunk 3 above)
Comment (eshen1991) [reply src/main/java/com/google/gcs/sdrs/controller/ServiceLayerException.java]: BaseController.errorResponse method relies on an exception type to generate proper error response. We can certainly use exception message to differentiate different type of errors. In this case, a new exception class works better as it'd not fundamentally how error response is generated

File: src/main/java/com/google/gcs/sdrs/dao/impl/RetentionRuleDaoImpl.java (hunk 4 above)
Comment (eshen1991) [reply src/main/java/com/google/gcs/sdrs/dao/impl/RetentionRuleDaoImpl.java]: done

File: src/main/java/com/google/gcs/sdrs/service/impl/RetentionRulesServiceImpl.java (hunk 5 above)
Comment (eshen1991) [reply src/main/java/com/google/gcs/sdrs/service/impl/RetentionRulesServiceImpl.java]: done

File: src/main/java/com/google/gcs/sdrs/service/impl/RetentionRulesServiceImpl.java (hunk 2 above)
Comment (eshen1991) [reply src/main/java/com/google/gcs/sdrs/service/impl/RetentionRulesServiceImpl.java]: done

**Pull request merged**